marlowe-test/test/Spec/Marlowe/StaticAnalysis/Regression.hs

```diff
@@ -0,0 +1,213 @@
+isTransactionNonPositiveDeposit (TransactionNonPositiveDeposit{}) = True
+isTransactionNonPositiveDeposit _ = False
+
+isTransactionShadowing :: TransactionWarning -> Bool
+isTransactionShadowing (TransactionShadowing{}) = True
+isTransactionShadowing _ = False
+
+isTransactionAssertionFailed :: TransactionWarning -> Bool
+isTransactionAssertionFailed (TransactionAssertionFailed{}) = True
+isTransactionAssertionFailed _ = False
+
+getWarning :: Contract -> IO (Maybe TransactionWarning)
+getWarning contract =
+  warningsTrace contract <&> \res -> do
+    (_, _, t) <- join $ hush res
+    listToMaybe t
+
+analysisWorks :: IO ()
+analysisWorks = do
+  let contract n d =
+        If
+          (DivValue (Constant n) (Constant d) `ValueGE` Constant 5)
+          Close
+          (Pay alicePk (Party alicePk) ada (Constant (-100)) Close)
+  result <- warningsTrace (contract 10 11)
+  assertBool "Analyse a contract" $ isRight result
+
+emptyTrace :: IO ()
+emptyTrace = do
+  let contract = Close
+  result <- warningsTrace contract
+  assertBool "Empty trace" $ isRight result && isNothing (fromRight Nothing result)
+
+nonPositivePay :: IO ()
+nonPositivePay = do
+  let contract = Pay alicePk (Party alicePk) ada (Constant (-100)) Close
+  result <- getWarning contract
+  assertBool "Detect negative pay" $ maybe False isTransactionNonPositivePay result
+  let contract2 = Pay alicePk (Party alicePk) ada (Constant 0) Close
+  result2 <- getWarning contract2
+  assertBool "Detect zero pay" $ maybe False isTransactionNonPositivePay result2
+
+partialPay :: IO ()
+partialPay = do
+  let contract = Pay alicePk (Party alicePk) ada (Constant 100) Close
+  result <- getWarning contract
+  assertBool "Detect partial pay" $ maybe False isTransactionPartialPay result
+
+nonPositiveDeposit :: IO ()
+nonPositiveDeposit = do
+  let contract v =
+        When
+          [ Case
+              ( Deposit
+                  alicePk
+                  alicePk
+                  (Token "" "")
+                  (Constant v)
+              )
+              Close
+          ]
+          1699974289397
+          Close
+  result <- getWarning (contract (-100))
+  assertBool "Negative deposit" $ maybe False isTransactionNonPositiveDeposit result
+  result2 <- getWarning (contract 0)
+  assertBool "Zero deposit" $ maybe False isTransactionNonPositiveDeposit result2
+
+transactionShadowing :: IO ()
+transactionShadowing = do
+  let contract = Let "x" (Constant 1) (Let "x" (Constant 2) Close)
+  result <- getWarning contract
+  assertBool "Shadowing x" $ maybe False isTransactionShadowing result
+
+assertionFailed :: IO ()
+assertionFailed = do
+  let contract = Let "x" (Constant 1) (Assert (ValueEQ (UseValue "x") (Constant 2)) Close)
+  result <- getWarning contract
+  assertBool "Detect partial pay" $ maybe False isTransactionAssertionFailed result
+
+-- The current UI drops in instrumentation in the form of assertions and then checks for AssertionFailed
```

Comment:
That reminds me there is a way of calling the static analysis `warningsTraceCustom` that takes a flag that allows you to only check for assertions (ignore warnings). And also there is a way to start with a particular state if you want, that is also accessible from `warningsTraceWithState`. The most generic one is `warningsTraceCustom`. We are not currently using the state functionality (or maybe we are for ensuring that minTime is at least 0), but we are using the ignore warnings bit for reachability (I think).

Reply:
We should record this documentation somewhere :) I saw these functions as I went through it, but there were no uses of some of them, so it was hard to tell what they were for.